### ISSUE TITLE
Add Safes, remove Safes and switching between them

### ISSUE THREAD

**Issue by tschubotz** [gnosis/safe#285]
# What is this feature about? (1 sentence)
This allows a user to add one or more existing Safes to their mobile app and switch between them.

# Why is it needed? What is the value? For whom do we build it?
Only after adding a Safe can the user use the rest of the app. Also, we expect that users own multiple Safes, so we would like to support that.

# High-level overview of the feature
- There is a list of Safes (Can be empty)
- User can add one or several Safes (one after each other)
  - (Edited the next few bullet points on June 17 to remove ambiguity)
  - By scanning a QR code
  - By pasting in an address into the field (by tapping the sheet action on here https://zpl.io/VD0nzY3)
  - By manually typing in an ENS name (into the field on https://zpl.io/2yvxdEo)
  - By pasting an ENS name (via long tap, iOS/Android native pasting functionality on https://zpl.io/2yvxdEo)
- User cannot add the same Safe twice (See [comment](https://github.com/gnosis/safe/issues/285#issuecomment-621691981) below)
- Entered address is validated if it's actually a Safe or not.
  - User can only proceed if it's a Safe address.
  - (Edit June 15) User can only proceed if mastercopy if >=1.0.0 (since web only supports this) and is one of the supported master copies of our backend.
- User needs to enter a (local) Safe name for that Safe
- User can change the local Safe name
- User can view the Safe
  - ENS name
  - Address
  - QR code
  - Identicon
- User can remove a Safe from their device
- User can switch between multiple Safes.
  - Safes are sorted in the order in which they were added. 
  - Active Safe is at the top. 
  - When you select another Safe in the list, the order does not change right away, but only when you open the switch safe screen (bottom sheet) again - then this selected Safe is at the top. 
  - Newly added Safe becomes active Safe, and hence will be at the top.
  - (June 18) On iOS switch Safe animation can be different than in designs (cf. #67 #97)
- When a Safe is selected, I can see its identicon and its address on top of the screen to know it is selected.
- On a cold app start, the user would see the last selected Safe.

**Comment by jpalvarezl** [comment]
Just a comment on the ENS resolver. Should it be configurable? (not even sure if that is a possibility in the old safe app) Maybe keeping the user on the loop as to which resolver we are using is important? From a trust perspective.

**Comment by posthnikova** [comment]
> On a cold app start, the user would start on the Safe list always

Maybe remember the last active Safe or make one Safe default instead of choosing from the list every time?

**Comment by tschubotz** [comment]
> Just a comment on the ENS resolver. Should it be configurable? (not even sure if that is a possibility in the old safe app) Maybe keeping the user on the loop as to which resolver we are using is important? From a trust perspective.

@jpalvarezl 
Interesting thoughts. 
I would not make the ENS resolver configurable right now. I don't see this as super important. I like the idea about showing which resolver is currently used. Maybe info about the used service endpoints etc can be added as well 🤔 I'll think about this. I think this can be added to #291 


> > On a cold app start, the user would start on the Safe list always
> 
> Maybe remember the last active Safe or make one Safe default instead of choosing from the list every time?

@posthnikova 
I know we have it in the web app but I think introducing this "default" flag unnecessary. 
I think there are 2 options: Either show the last active one or always start from the list. I figured that always starting with the list is the best way without any implicit logic needed. What would you do?

**Comment by tschubotz** [comment]
I just added this item:

> - User needs to enter a (local) Safe name for that Safe
> - User can change the local Safe name

Since I think it's important to be able to differentiate Safes + the web interface has that functionality.

**Comment by posthnikova** [comment]
@tschubotz My preferred option would be to remember the last active Safe because this is the usual behaviour of apps with multiple accounts. If introducing logic is hard, let's start from the list.

**Comment by tschubotz** [comment]
> @tschubotz My preferred option would be to remember the last active Safe because this is the usual behaviour of apps with multiple accounts. If introducing logic is hard, let's start from the list.

You are right, e.g. email apps do that. Also the web interface opens on a Safe and not a list. I'll change it in the description.

**Comment by elgatovital** [comment]
we should also validate safe address before adding it

**Comment by tschubotz** [comment]
> we should also validate safe address before adding it

Added to description ✅

**Comment by jpalvarezl** [comment]
There is seeemingly no UI for removing safes from the list.

**Comment by tschubotz** [comment]
> There is seemingly no UI for removing safes from the list.

It's in the Safe settings on the very bottom :) 
https://zpl.io/2ZN0AAd

**Comment by jpalvarezl** [comment]
Ah thanks! Sorry I missed it.
Then I have another question: Should removing the safe be part of this epic then? Because as I understand it, UI-wise we have an empty state, a list from which you can define the active safe and input of new safes. This seems to belong then to some sort "settings" epic or something like that in which the associated data in the screen is different.

**Comment by tschubotz** [comment]
It logically belongs to the CRUD functions. That's why I put it in here. But it's on a different screen. I don't mind if we don't have the remove functionality, yet and only implement removing #288 
Let me ask in Slack to make sure this is cool with everyone.

**Comment by jpalvarezl** [comment]
I was not aware that you were segmenting by CRUD, sorry for bringing this up.

**Comment by tschubotz** [comment]
> I was not aware that you were segmenting by CRUD, sorry for bringing this up.

No need to apologize, please continue asking. We are currently figuring out our process right now, right? :)

**Comment by elgatovital** [comment]
currently, there is no check whether safe was already added to the device => add flow can be started for an already added safe, possibly replacing local name. We should inform the user that safe was already added and abort the flow. Users should be able to change safe name from safe settings anyway. What do you think? @tschubotz @posthnikova

**Comment by tschubotz** [comment]
@elgatovital absolutely agree, thanks for bringing this up. 

It should not be possible to add the same Safe twice. This is how iOS solved this currently. 
![IMG_6923362853E4-1](https://user-images.githubusercontent.com/1380529/80689195-3fb55a80-8acd-11ea-9cce-35b0e379ec48.jpeg)

I suggest doing the same on Android. What do you think?

**Comment by sche** [comment]
How about auto-populating Safe Name when adding Safe by ENS?
<img width="381" alt="Screenshot 2020-06-17 at 18 34 59" src="https://user-images.githubusercontent.com/2284385/84925026-a288a280-b0c9-11ea-83e2-b9cb356bb4ac.png">

**Comment by tschubotz** [comment]
> How about auto-populating Safe Name when adding Safe by ENS?

Nice idea however this is additional logic, I would rather add this as separate ticket so we can properly add this on both platforms. I'm scared that this is a first step towards feature creep :)

I create a ticket for this to keep track #331

**Comment by liliya-soroka** [comment]
The epic is closed as there are no critical or major issues, only minor and postponed
